presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkShufflePageInput.java

```diff
@@ -0,0 +1,178 @@
+    private int currentIteratorIndex;
+    @GuardedBy("this")
+    private final Stats stats = new Stats();
+
+    public PrestoSparkShufflePageInput(
+            List<Type> types,
+            List<PrestoSparkShuffleInput> shuffleInputs,
+            int taskId,
+            CollectionAccumulator<PrestoSparkShuffleStats> shuffleStatsCollector)
+    {
+        this.types = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+        this.shuffleInputs = ImmutableList.copyOf(requireNonNull(shuffleInputs, "shuffleInputs is null"));
+        this.taskId = taskId;
+        this.shuffleStatsCollector = requireNonNull(shuffleStatsCollector, "shuffleStatsCollector is null");
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        SliceOutput output = new DynamicSliceOutput(types.isEmpty() ? 0 : BUFFER_SIZE);
+        int rowCount = 0;
+        synchronized (this) {
+            while (currentIteratorIndex < shuffleInputs.size()) {
+                PrestoSparkShuffleInput input = shuffleInputs.get(currentIteratorIndex);
+                Iterator<Tuple2<MutablePartitionId, PrestoSparkMutableRow>> iterator = input.getIterator();
+                long processedBytes = 0;
+                long start = System.currentTimeMillis();
+                while (iterator.hasNext() && output.size() <= TARGET_SIZE && rowCount <= MAX_ROWS_PER_PAGE) {
+                    PrestoSparkMutableRow row = iterator.next()._2;
+                    if (row.getBuffer() != null) {
+                        ByteBuffer buffer = row.getBuffer();
+                        output.writeBytes(buffer.array(), buffer.arrayOffset() + buffer.position(), buffer.remaining());
+                        processedBytes += buffer.remaining();
+                    }
+                    else if (row.getArray() != null) {
+                        output.writeBytes(row.getArray(), row.getOffset(), row.getLength());
+                        processedBytes += row.getLength();
+                    }
+                    else {
+                        throw new IllegalArgumentException("Unexpected PrestoSparkMutableRow: 'buffer' and 'array' fields are both null");
+                    }
+                    rowCount++;
+                }
+                long end = System.currentTimeMillis();
+                stats.accumulate(rowCount, processedBytes, end - start);
+                if (!iterator.hasNext()) {
+                    PrestoSparkShuffleStats shuffleStats = new PrestoSparkShuffleStats(
+                            input.getFragmentId(),
+                            taskId,
+                            READ,
+                            stats.getProcessedRows(),
+                            stats.getProcessedBytes(),
+                            stats.getElapsedWallTimeMills());
+                    shuffleStatsCollector.add(shuffleStats);
+                    stats.reset();
+                    currentIteratorIndex++;
+                }
+                else {
```

Comment:
Why break here? Is it supposed to process next element in List<PrestoSparkShuffleInput> ?

Reply:
The condition of the upper loop (that reads the data) is: `iterator.hasNext() && output.size() <= TARGET_SIZE && rowCount <= MAX_ROWS_PER_PAGE` meaning "keep reading data until the iterator is not empty and the threshold per page is not crossed". If the upper loop finishes it means either of two conditions:

1. The current iterator has no more data
2. The buffer is full

For the first condition we transition to the second iterator. And while we do that - we log statistics for the first (there's 1 iterator per input fragment).

For the second condition we need to end the loop, as we've read the whole page.